go/types/simplify.go

```diff
@@ -208,8 +212,8 @@ func simplifyStructs(expectedName string, ts typeset) *Type {
 	for name, fti := range allFields {
 		fields = append(fields, StructField{
 			Name:     name,
-			Type:     makeSimplifiedTypeImpl(fti.typeset),
-			Optional: fti.count < count,
+			Type:     makeSimplifiedTypeImpl(fti.typeset, intersectStructs),
+			Optional: !(intersectStructs && fti.anyRequired) && fti.count < count,
```

Comment:
While a little more verbose, this would be slightly easier to read if anyRequired were instead named anyNonOptional. This would make the relationship to 'optional' immediately obvious.